### PR TITLE
Upload to gdrive on drag&drop event

### PR DESCRIPTION
**Pull request by denar90**
Hey 👋 
I've taped some video 
https://monosnap.com/file/bt6S1CD2NXciib838DBaFEuq2i72Kn
about how it works. Later we can make gif and add on a main page :)

Btw I dunno if we need to add some styles when user drag&drop file on main page. wdyt?

### CODE REVIEW

**Comment by paulirish** [comment]
Yup! i like the functionality. I think we need some UX to finish this.

When I've dragged a trace into here, i've wanted one of two things

1) just view the trace in the viewer. it's a "viewer" after all. :)
2) view it and create a permalink so other people can view it as well.

I think the ideal UX is as follows:

1. on dragover you get the normal drag drop affordances. (we can just call `initializeDevTools` and `makeDevToolsVisible` and then let the timeline/perf panel's existing hover stuff take over
2. on drop, the panel loads the trace
3. as a user, i choose to upload the trace to drive or not. 

so we'll need some UX for #3. i guess something up in the top right next to our offline status?

**Comment by denar90** [comment]
Got your point. Really interesting UX, I like it 👍

**Comment by denar90** [comment]
@paulirish are there any method to get uploaded timeline data?

I see 2 ways.

----
 If I got it right, after drag/dropping, trace is saved in tempFile. 
So we can to read this file with data and send read stream to gDrive (kinda save to file).

---
Second way is to get data using `Timeline.TimelinePanel.instance()._model.records()` but I dunno how to convert array of `TimelineModel.TimelineModel.Record` data into json format...

Which way would you suggest?

**Comment by paulirish** [comment]
let me know if you're getting blocked here and I can take a look

**Comment by paulirish** [comment]
btw take a look at https://github.com/ChromeDevTools/timeline-viewer/blob/master/docs/timeline_viewer.js#L116 as I think we might be receiving it right here already.. :)

**Comment by denar90** [comment]
Thanks, sorry, but I didn't have time to look at it yet. Gimme couple days, a lot of stuff going around :)

File: docs/timeline_viewer.js [hunk 1]
```diff
@@ -262,56 +366,172 @@ class Viewer {
...
-      xhr.open('GET', url);
+      const xhr = new XMLHttpRequest();
+      xhr.open(method, url);
       callbetween && callbetween(xhr);
-      xhr.onprogress = this.updateProgress.bind(this);
-      xhr.onload = _ => resolve(xhr.responseText);
-      xhr.onerror = (err => {
+      //show progress only while getting data
+      if (method === 'GET') {
+        xhr.onprogress = this.updateProgress.bind(this);
```
Comment (denar90) [inline docs/timeline_viewer.js]: While sending trace to drive progress is shown but I dunno how to hide it. So tmp fix for now

**Comment by denar90** [comment]
@paulirish Here new video about new functionality https://monosnap.com/file/0MvUZy9NWl5quUQG0Lioyu2udnt70n

There some issue about hidden panel tabs, it's because of loading resources order...

P.S. Monkey patching was fun :) I got a lot about how dev tools work.

File: docs/timeline_viewer.js [hunk 2]
```diff
@@ -111,17 +183,35 @@ class Viewer {
...
-      finishWriting: _ => { }
+      write: function(strings) {
+        this._chunks = this._chunks.concat(strings);
+      },
+      remove: function() {
+        this._file = null;
+        this._chunks = [];
+      },
+      finishWriting: function() {
+        this._file = new Blob(this._chunks.filter(Object), { type: "text/plain" });
```
Comment (denar90) [inline docs/timeline_viewer.js]: Kinda file emulation stuff.

**Comment by denar90** [comment]
So, it's done...
..for now :)

**Comment by paulirish** [comment]
cool will look this weekend!
​

**Comment by denar90** [comment]
💃  thx :)

File: docs/timeline_viewer.js [hunk 3]
```diff
@@ -373,6 +487,108 @@ class Viewer {
...
+      const handleDrop = dropTarget._handleDrop;
+      dropTarget._handleDrop = function(dataTransfer) {
+        viewer.toggleUploadToDriveElem(viewer.canUploadToDrive);
+        handleDrop.apply(dropTarget, arguments);
+      };
+    }
+  }
+
+  uploadTimelineData() {
+    Timeline.TimelinePanel.instance()._backingStorage._file.read(event => {
```
Comment (denar90) [inline docs/timeline_viewer.js]: try read from stream stuff
```
+   * @return {?Common.OutputStream}
+   */
+  _saveDataStream() {
+    if (this._state !== Timeline.TimelinePanel.State.Idle)
+      return null;
+    if (this._model.isEmpty())
+      return null;
+
+    var stream = new Common.StringOutputStream;
+    var saver = new Timeline.TracingTimelineSaver();
+    this._tracingModelBackingStorage.writeToStream(stream, saver);
+    return stream;
+  }

```

Comment (denar90) [inline docs/timeline_viewer.js]: No, we can't because of https://github.com/ChromeDevTools/timeline-viewer/blob/master/docs/timeline_viewer.js#L120-L126

**Comment by paulirish** [comment]
merged along with #25 landing

**Pull request closed without merging**